Update WebXR mode to match current immersive-ar

klausw opened the pull request:
Get WebXR mode working again using the current API as implemented in Chrome Canary using immersive-ar mode. 

This commit is a snapshot of an airplane hacking session, it's in rough shape due to having a barely-functioning dev environment. Please ignore superficial issues such as indentation, those should obviously be fixed before merging. There's also extra logging, and I was taking shortcuts to get it working such as overriding feature checks. I'm turning it into a PR at @elalish 's suggestion, at least it should be useful as a roadmap of areas that need changing.

FYI, the hit test API is likely to undergo more changes in the near future again, so that will need further tweaks at that time.

Fixes #495

klausw commented:
Deployed version is at https://klausw.github.io/model-viewer/www/examples/augmented-reality.html

cdata commented:
This is exciting 👍 nice to see some help in this area.  Thanks @klausw !

elalish commented:
@klausw Do you want to keep working on this PR, or should I just pull it and finish it on my side?

klausw commented:
Please review, I've rebased it and cleaned up the changes, including removing the no-longer-needed fullscreen mode workarounds and reusing the existing threejs WebGLRenderer instead of creating a new one. As a side effect, that made the regular Astronaut work, so I no longer need to switch to the unlit variant.

klausw commented:
FYI, updated test URL is https://klausw.github.io/model-viewer/examples/augmented-reality.html

elalish commented on src/three-components/ARRenderer.ts:
```diff
@@ -232,18 +186,20 @@ export class ARRenderer extends EventDispatcher {
 
   [$postSessionCleanup]() {
+    (this.renderer as any).setFramebuffer(null);
+    this.parentRenderer.setRendererSize(1, 1);
```
What is this for?

klausw replied:
I've added a comment, this triggers a resize in the render loop. The parent Renderer code does this also, I was following its example, but I'm not sure if this is the best way to do so.

elalish commented on src/three-components/ARRenderer.ts:
```diff
@@ -377,12 +314,13 @@ height: 100%;`);
+      const viewport =
+          (session.renderState.baseLayer as XRWebGLLayer).getViewport(view);
+      this.renderer.setViewport(0, 0, viewport.width, viewport.height);
+      this.renderer.setSize(viewport.width, viewport.height, false);
```
setSize seems weird here. Are there ever multiple views? If there were I don't think the size would be the same as the viewport.

klausw replied:
The "const viewport" assignment above specifically gets the viewport size for use with the current view, either the whole framebuffer or an appropriate subsection. See https://immersive-web.github.io/webxr/#dom-xrwebgllayer-getviewport :
> getViewport() queries the XRViewport the given XRView should use when rendering to the layer.

In this case only one view is expected, but the view loop should work for the more general case also.

klausw replied:
Sorry, I had described the way it should work, but the implementation didn't match. Should be fixed by https://github.com/GoogleWebComponents/model-viewer/pull/757/commits/311706df0e21924830be4c5c0a7ff18e8ed6171d

elalish commented on src/types/webxr.d.ts:
```diff
@@ -109,25 +106,30 @@ type XRTargetRayMode = 'gaze'|'tracked-pointer'|'screen';
+
+interface XRSession extends EventTarget {
+  renderState: XRRenderState;
+  updateRenderState(obj: any): any;
```
Is this really all the typing we have for this method?

klausw replied:
I've added more specific type info using XRRenderStateInit.

elalish commented on src/three-components/ARRenderer.ts:
```diff
@@ -52,93 +47,66 @@ export class ARRenderer extends EventDispatcher {
-    this.renderer = new WebGLRenderer(
-        {canvas: this.inputCanvas, context: this.inputContext});
-    this.renderer.setSize(window.innerWidth, window.innerHeight);
     this.renderer.setPixelRatio(1);
```
If we're going to reuse our renderer, we'll need to reset this to the device DPR when we're done.

klausw replied:
I added a comment in postSessionCleanup. Renderer.render auto-adjusts the device pixel ratio as needed already.

elalish commented on src/constants.ts:
```diff
@@ -17,21 +17,16 @@
     HAS_WEBXR_DEVICE_API && self.XRSession!.prototype.requestHitTest;
 
-export const HAS_FULLSCREEN_API = document.documentElement != null &&
-    document.documentElement.requestFullscreen != null;
```
I think there's some more fullscreen stuff that hasn't been reverted yet.

klausw replied:
This constant only appeared to be used for WebXR mode. It was used in utilities.ts's assertIsArCandidate which is only used in the WebXR ARRenderer.ts, and I had removed it there since immersive-ar mode doesn't require use of the Fullscreen API. (I think that was part of the reason I was under the impression that the other AR code didn't need fullscreen.)

klausw commented on src/features/ar.ts:
```diff
@@ -153,32 +140,9 @@ export const ARMixin = <T extends Constructor<ModelViewerElementBase>>(
-    // As of m76, intents could no longer be triggered successfully if they
-    // were dispatched in the same handler as the fullscreen request. The
-    // workaround is to split both effects into their own event handlers.
-    // @see https://github.com/GoogleWebComponents/model-viewer/issues/693
```
I've restored that code in https://github.com/GoogleWebComponents/model-viewer/pull/757/commits/45c4a9b132ffee949c67f998f565a3ab57397474 , I had misunderstood how this works and didn't realize that AR_VIEWER mode also used requestFullscreen.

klausw commented on src/three-components/ARRenderer.ts:
```diff
@@ -52,93 +47,66 @@ export class ARRenderer extends EventDispatcher {
 
-  constructor(
-      inputCanvas: HTMLCanvasElement, inputContext: WebGLRenderingContext) {
+  constructor(parentRenderer: Renderer) {
```
I've added some clarifying comments, and replaced the fromInlineRenderer call with using the constructor directly, both here and in the ARRenderer-spec test.

klausw commented on src/three-components/ARRenderer.ts:
```diff
@@ -52,93 +47,66 @@ export class ARRenderer extends EventDispatcher {
+    await waitForAnimationFrame;
 
+    (this.renderer as any)
+        .setFramebuffer(
```
Done.

klausw commented:
PTAL, and I've updated the gh-pages at  https://klausw.github.io/model-viewer/examples/augmented-reality.html , though github is sometimes slow in consistently serving that.

elalish commented:
Thanks, this is looking great! I want to do a bit more testing before merging, but that's it.

elalish commented:
@klausw Would it be possible to add an extra WebXR button in this PR? When I test I can only launch SceneViewer, unless I'm missing something?

klausw commented:
The button should currently switch to WebXR mode if the prerequisites are
met - you need to enable WebXR Device API and WebXR AR Module in
chrome://flags, and use a reasonably current Chrome Canary on an
ARCore-supported phone with Android >= 8.0 / Oreo.

I'll look into adding a separate button.

On Fri, Sep 27, 2019, 10:28 Emmett Lalish <notifications@github.com> wrote:

> @klausw <https://github.com/klausw> Would it be possible to add an extra
> WebXR button in this PR? When I test I can only launch SceneViewer, unless
> I'm missing something?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleWebComponents/model-viewer/pull/757?email_source=notifications&email_token=AAKT5X6MPTZQ7CUXU55X4UDQLY7DPA5CNFSM4IXIMPT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7ZSRSQ#issuecomment-536029386>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKT5X63SVUBFWPBWAV76BTQLY7DPANCNFSM4IXIMPTQ>
> .
>

klausw commented:
On Fri, Sep 27, 2019 at 10:55 AM Klaus Weidner <notifications@github.com>
wrote:

> The button should currently switch to WebXR mode if the prerequisites are
> met - you need to enable WebXR Device API and WebXR AR Module in
> chrome://flags, and use a reasonably current Chrome Canary on an
> ARCore-supported phone with Android >= 8.0 / Oreo.
>

Correction, it also needs WebXR Hit Test API enabled in chrome://flags.
This also works on the stable Chrome release 77 with the same flag changes,
except that this version doesn't have a "WebXR AR module" flag. (It's
auto-activated by enabling hit test).

klausw commented:
> The main thing is to put back the fullscreen stuff so that SceneViewer still works. We'll have to test it manually on a Pixel 3.

I had reverted those changes to restore the SceneViewer fullscreen path, AFAIK the changes should now only affect WebXR mode. I've tested this by disabling WebXR AR Module (removing one of the other prerequisites should also work), and verifying that the AR button then launches Scene Viewer.

(Note to self and other testers: scene viewer does NOT appear to work for a port-forwarded localhost mode such as "npm run dev" using localhost:8000. It gets launched but can't load the model. It should work for the github-hosted version at https://klausw.github.io/model-viewer/examples/augmented-reality.html .)
